policy_agent/policy_agent.py

```diff
@@ -0,0 +1,643 @@
+#!/usr/bin/python
+import os
+import sys
+import time
+import json
+import Queue
+import logging
+import requests
+import pycalico
+
+from threading import Thread
+from subprocess import check_output
+from contextlib import closing
+from pycalico.datastore_datatypes import Rules, Rule
+from pycalico.datastore_errors import ProfileNotInEndpoint, ProfileAlreadyInEndpoint
+from pycalico.datastore import DatastoreClient, PROFILE_PATH
+
+KUBE_API_ROOT = os.environ.get('KUBE_API_ROOT',
+                               'http://localhost:8080/api/v1/')
+
+ANNOTATION_NAMESPACE = "projectcalico.org/"
+EPID_ANNOTATION_KEY = "%sendpointID" % ANNOTATION_NAMESPACE
+
+POLICY_LOG_DIR = "/var/log/calico/policy"
+POLICY_LOG = "%s/calico.log" % POLICY_LOG_DIR
+
+LOG_FORMAT = '%(asctime)s %(process)d %(levelname)s %(filename)s: %(message)s'
+LOG_LEVEL = os.environ.get('LOG_LEVEL', 'INFO').upper()
+
+_log = logging.getLogger(__name__)
+_datastore_client = DatastoreClient()
+
+
+class PolicyAgent():
+
+    """
+    The Policy Agent is responsible for maintaining Watch Threads/Queues and internal resource lists
+    """
+
+    def __init__(self):
+        self.q = Queue.Queue()
+        self.namespaces = dict()
+        self.services = dict()
+        self.pods = dict()
+        self.endpoints = dict()
+
+        self.changed_namespaces = dict()
+        self.changed_services = dict()
+        self.changed_pods = dict()
+        self.changed_endpoints = dict()
+
+    def run(self):
+        """
+        PolicyAgent.run() is called at program init to spawn watch threads and parse their responses
+        """
+        self.init_lists()
+        self.init_threads()
+
+        self.PodWatcher.start()
+        self.SvcWatcher.start()
+        self.NsWatcher.start()
+        self.EptsWatcher.start()
+
+        while True:
+            self.read_responses()
+
+    def init_lists(self):
+        _log.info("Initializing Resource Lists")
+        initNamespaces = _get_api_list("namespaces")
+        initServices = _get_api_list("services")
+        initPods = _get_api_list("pods")
+        initEndpoints = _get_api_list("endpoints")
+
+        for ns in initNamespaces["items"]:
+            self.process_resource(action="ADDED", kind="Namespace", target=ns)
+
+        for svc in initServices["items"]:
+            self.process_resource(action="ADDED", kind="Service", target=svc)
+
+        for pod in initPods["items"]:
+            self.process_resource(action="ADDED", kind="Pod", target=pod)
+
+        for ep in initEndpoints["items"]:
+            self.process_resource(action="ADDED", kind="Endpoints", target=ep)
+
+        self.nsRV = initEndpoints["metadata"]["resourceVersion"]
```

Comment:
No need to abbreviate so much - make the code harder to figure out.

`nsRV` -> `nsResourceVersion`, etc.